python/mlx/nn/init.py

```diff
@@ -348,3 +348,56 @@ def initializer(
         return mx.random.uniform(-limit, limit, a.shape, dtype=dtype)
 
     return initializer
+
+
+def sparse(
+    mean: float = 0.0,
+    std: float = 1.0,
+    sparsity: float = None,
+    dtype: mx.Dtype = mx.float32,
+) -> Callable[[mx.array], mx.array]:
+    r"""An initializer that returns a sparse matrix.
+
+    Args:
+        mean (float, optional): Mean of the normal distribution. Default:
+          ``0.0``.
+        std (float, optional): Standard deviation of the normal distribution.
+          Default: ``1.0``.
+        sparsity (float, optional): The fraction of elements in each column to
+          be set to zero. Default: ``None``.
+        dtype (Dtype, optional): The data type of the array. Default:
+          ``float32``.
+
+    Returns:
+        Callable[[array], array]: An initializer that returns an array with the
+        same shape as the input, filled with samples from a normal distribution.
+
+    Example:
+
+        >>> init_fn = nn.init.sparse(sparsity=0.5)
+        >>> init_fn(mx.zeros((2, 2)))
+        array([[-1.91187, -0.117483],
+       [0, 0]], dtype=float32)
+    """
+
+    def initializer(a: mx.array) -> mx.array:
+
+        if sparsity == None:
+            raise ValueError(f"Sparsity argument cannot be {sparsity}")
+
+        if a.ndim != 2:
```

Comment:
What is preventing us from initializing all arrays here? I think this check should be removed.

Reply:
@angeloskath, I believe that `mx.random.Bernoulli` might generate matrices where entire columns consist solely of 1s or 0s. Our approach, however, emphasizes a more even distribution of sparsity across dimensions, specifically the columns in this case. Sparse initialization is primarily used for initializing 2D weights, and ensuring an even distribution of sparsity becomes increasingly complex as the number of dimensions grows. Therefore, we opted to stick with the current implementation.

Reply:
Still the loop is a significantly inefficient way of implementing this. We will have thousands of columns and hundreds of arrays. A more efficient implementation exactly equivalent to your original one is the following:

```python
order = mx.argsort(mx.random.uniform(shape=(a.shape)), axis=0)
a = mx.random.normal(shape=a.shape, scale=std, loc=mean, dtype=dtype)
a[order[:num_zeros], mx.arange(cols)[None]] = 0
```

Another detail is that you probably want to initialize each row to be partially 0 instead of each column. In all linear layers we do `x @ w.T` so having a completely 0 row is probably the thing you are trying to avoid.

Reply:
@angeloskath Refactored the code with the suggested changes. Could you please review it?